azurelinuxagent/common/logcollector.py

```diff
@@ -0,0 +1,360 @@
+# Microsoft Azure Linux Agent
+#
+# Copyright 2020 Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+#
+
+from datetime import datetime
+import glob
+from heapq import heappush, heappop
+import os
+import tarfile
+import zipfile
+
+from azurelinuxagent.common.future import ustr
+from azurelinuxagent.common.utils.fileutil import mkdir, read_file, rm_files, append_file, append_items_to_file
+from azurelinuxagent.common.utils.shellutil import run_command
+from azurelinuxagent.common.utils.textutil import safe_shlex_split
+
+
+LOG_COLLECTOR_DIR = '/var/lib/waagent/logcollector'
+TRUNCATED_FILES_DIR = '/var/truncated'
+
+OUTPUT_RESULTS_FILE_PATH = os.path.join(LOG_COLLECTOR_DIR, "results.txt")
+OUTPUT_ARCHIVE_PATH = os.path.join(LOG_COLLECTOR_DIR, "logs.tar")
+COMPRESSED_ARCHIVE_PATH = os.path.join(LOG_COLLECTOR_DIR, "logs.zip")
+
+MUST_COLLECT_FILES = [
+    '/var/log/waagent*',
+    '/var/lib/waagent/GoalState.*.xml',
+    '/var/lib/waagent/waagent_status.json',
+    '/var/lib/waagent/history/*.zip',
+    '/var/log/azure/*/*',
+    '/var/log/azure/*/*/*',
+    '/var/lib/waagent/ExtensionsConfig.*.xml',
+    '/var/lib/waagent/HostingEnvironmentConfig.xml',
+    '/var/lib/waagent/error.json'
+]
+
+FILE_SIZE_LIMIT = 30 * 1024 * 1024  # 30 MB
+ARCHIVE_SIZE_LIMIT = 150 * 1024 * 1024  # 150 MB
+
+
+class LogCollector(object):
+
+    def __init__(self, manifest_file_path):
+        self.manifest_file_path = manifest_file_path
+        self.must_collect_files = self._expand_must_collect_files()
+
+    def _expand_must_collect_files(self):
+        # Match the regexes from the MUST_COLLECT_FILES list to existing file paths on disk.
+        manifest = []
+        for path in MUST_COLLECT_FILES:
+            expanded_paths = self._expand_path(path)
+            manifest.extend(expanded_paths)
+
+        return manifest
+
+    def _read_manifest_file(self):
+        return read_file(self.manifest_file_path).splitlines()
+
+    @staticmethod
+    def _log_to_results_file(entry):
```

Comment:
why not use python's logging facilities for this (e.g. errors should be clearly marked)

Reply:
Didn't occur to me to use that module, honestly. I was mimicking the output of IID and their results file which is just a simple outline of operations performed. I don't think there's added value in using Python's logger in this case.

Reply:
k

Reply:
It might be worth checking out the logging library though mainly because it gives you a lot of additional functionalities for free. 
For example I noticed that you're not logging the timestamp to the `results_file` which would be really helpful during debugging if something goes wrong. You can always add it here but it might be a simpler change to do it if using a module vs modifying the custom implementation.

Reply:
Added.